File: kontiva-macos/Sources/KontivaCore/Insights.swift
import Foundation

/// How prominent an insight is. Drives ordering, colour, and icon in the UI.
public enum InsightSeverity: String, Sendable, Equatable {
    case warning   // needs attention (overspending, overdue bills)
    case tip       // an opportunity to improve
    case info      // a neutral observation about where money goes
    case positive  // something going well

    var priority: Int {
        switch self {
        case .warning: return 0
        case .tip:     return 1
        case .info:    return 2
        case .positive:return 3
        }
    }
}

/// A single budget insight. Pure data with the numbers already computed — the UI
/// turns each case into a localized title + detail (see `KontivaUI`).
public enum Insight: Equatable, Sendable, Identifiable {
    case overspending(deficit: Money)
    case tightBudget(available: Money, pct: Int)
    case healthySurplus(available: Money, pct: Int)
    case highFixedBurden(total: Money, pct: Int)
    case highHousing(amount: Money, pct: Int)
    case largestFixedCost(name: String, amount: Money, pctOfFixed: Int)
    case largestVariable(name: String, amount: Money)
    case overdueBills(count: Int, total: Money)
    case noSavings
    case goodSavingsRate(monthly: Money, pct: Int)
    case allHealthy

    public var id: String {
        switch self {
        case .overspending:    return "overspending"
        case .tightBudget:     return "tightBudget"
        case .healthySurplus:  return "healthySurplus"
        case .highFixedBurden: return "highFixedBurden"
        case .highHousing:     return "highHousing"
        case .largestFixedCost:return "largestFixedCost"
        case .largestVariable: return "largestVariable"
        case .overdueBills:    return "overdueBills"
        case .noSavings:       return "noSavings"
        case .goodSavingsRate: return "goodSavingsRate"
        case .allHealthy:      return "allHealthy"
        }
    }

    public var severity: InsightSeverity {
        switch self {
        case .overspending, .overdueBills:       return .warning
        case .highFixedBurden, .highHousing,
             .noSavings, .tightBudget:           return .tip
        case .largestFixedCost, .largestVariable:return .info
        case .healthySurplus, .goodSavingsRate,
             .allHealthy:                        return .positive
        }
    }
}

/// Rule-based analysis of the current budget snapshot. Swiss-flavoured thresholds
/// (e.g. housing ≤ ~⅓ of income, save ≥ ~10%). No history is required; this
/// describes *this month's plan*. Month-over-month comparison is a future addition
/// that needs stored history.
public enum InsightEngine {

    public static func analyze(
        incomes: [Income],
        fixedCosts: [RecurringFixedExpense],
        variableBudgets: [VariableMonthlyBudget],
        bills: [OneOffBill],
        savingsGoals: [SavingsGoal],
        availability: MonthlyAvailability,
        asOf reference: Date = Date(),
        calendar: Calendar = .swiss
    ) -> [Insight] {

        var insights: [Insight] = []
        let income = availability.netIncomeThisMonth
        let available = availability.available

        // Available balance health.
        if available.isNegative {
            insights.append(.overspending(deficit: Money(rappen: -available.rappen)))
        } else if income.isPositive {
            let pct = available.percent(of: income)
            if pct < 10 {
                insights.append(.tightBudget(available: available, pct: pct))
            } else if pct >= 20 {
                insights.append(.healthySurplus(available: available, pct: pct))
            }
        }

        // Only fixed costs active this month are considered (a finished standing
        // order no longer burdens the budget).
        let activeFixed = fixedCosts.filter { $0.isActive(in: reference, calendar: calendar) }

        // Fixed-cost burden.
        let totalFixed = activeFixed.map(\.monthlyAmount).total()
        if income.isPositive && totalFixed.isPositive {
            let pct = totalFixed.percent(of: income)
            if pct >= 50 { insights.append(.highFixedBurden(total: totalFixed, pct: pct)) }
        }

        // Housing ratio (rent-category fixed costs vs income; Swiss guideline ~⅓).
        let rent = activeFixed.filter { $0.category == .rent }.map(\.monthlyAmount).total()
        if rent.isPositive && income.isPositive {
            let pct = rent.percent(of: income)
            if pct >= 33 { insights.append(.highHousing(amount: rent, pct: pct)) }
        }

        // Largest fixed cost (only worth noting when there are several).
        if activeFixed.count >= 2,
           let largest = activeFixed.max(by: { $0.monthlyAmount < $1.monthlyAmount }),
           largest.monthlyAmount.isPositive {
            let pctOfFixed = largest.monthlyAmount.percent(of: totalFixed)
            insights.append(.largestFixedCost(name: largest.name, amount: largest.monthlyAmount,
                                              pctOfFixed: pctOfFixed))
        }

        // Largest variable budget.
        if variableBudgets.count >= 2,
           let largest = variableBudgets.max(by: { $0.plannedAmount < $1.plannedAmount }),
           largest.plannedAmount.isPositive {
            insights.append(.largestVariable(name: largest.name, amount: largest.plannedAmount))
        }

        // Overdue bills.
        let overdue = bills.filter {
            BillClassifier.state(of: $0, asOf: reference, calendar: calendar) == .overdue
        }
        if !overdue.isEmpty {
            insights.append(.overdueBills(count: overdue.count, total: overdue.map(\.amount).total()))
        }

        // Savings.
        let monthlySavings = savingsGoals.compactMap(\.monthlyContribution).total()
        if income.isPositive {
            if monthlySavings.isZero && available.isPositive && !available.isZero {
                insights.append(.noSavings)
            } else if monthlySavings.isPositive {
                let pct = monthlySavings.percent(of: income)
                if pct >= 10 { insights.append(.goodSavingsRate(monthly: monthlySavings, pct: pct)) }
            }
        }

        if insights.isEmpty { insights.append(.allHealthy) }

        // Most important first; stable within a severity.
        return insights.enumerated()
            .sorted { ($0.element.severity.priority, $0.offset) < ($1.element.severity.priority, $1.offset) }
            .map(\.element)
    }
}

File: kontiva-android/app/src/main/java/ch/kontiva/android/core/Insights.kt
package ch.kontiva.android.core

import ch.kontiva.android.core.l10n.L10nKey
import java.time.LocalDate

enum class InsightSeverity(val priority: Int) { WARNING(0), TIP(1), INFO(2), POSITIVE(3) }

/** A single budget insight with its numbers; the UI turns it into a localized card. */
sealed class Insight(val severity: InsightSeverity, val titleKey: L10nKey) {
    data class Overspending(val deficit: Money) : Insight(InsightSeverity.WARNING, L10nKey.insightOverspending)
    data class TightBudget(val available: Money, val pct: Int) : Insight(InsightSeverity.TIP, L10nKey.insightTightBudget)
    data class HealthySurplus(val available: Money, val pct: Int) : Insight(InsightSeverity.POSITIVE, L10nKey.insightHealthySurplus)
    data class HighFixedBurden(val total: Money, val pct: Int) : Insight(InsightSeverity.TIP, L10nKey.insightHighFixed)
    data class HighHousing(val amount: Money, val pct: Int) : Insight(InsightSeverity.TIP, L10nKey.insightHighHousing)
    data class LargestFixedCost(val name: String, val amount: Money, val pctOfFixed: Int) : Insight(InsightSeverity.INFO, L10nKey.insightLargestFixed)
    data class LargestVariable(val name: String, val amount: Money) : Insight(InsightSeverity.INFO, L10nKey.insightLargestVariable)
    data class OverdueBills(val count: Int, val total: Money) : Insight(InsightSeverity.WARNING, L10nKey.insightOverdue)
    data object NoSavings : Insight(InsightSeverity.TIP, L10nKey.insightNoSavings)
    data class GoodSavingsRate(val monthly: Money, val pct: Int) : Insight(InsightSeverity.POSITIVE, L10nKey.insightGoodSavings)
    data object AllHealthy : Insight(InsightSeverity.POSITIVE, L10nKey.insightAllHealthy)
}

/** Rule-based analysis of this month's plan (Swiss thresholds). 1:1 with iOS. */
object InsightEngine {
    fun analyze(
        fixedCosts: List<RecurringFixedExpense>,
        variableBudgets: List<VariableMonthlyBudget>,
        bills: List<OneOffBill>,
        savingsGoals: List<SavingsGoal>,
        availability: MonthlyAvailability,
        today: LocalDate = LocalDate.now(),
    ): List<Insight> {
        val out = mutableListOf<Insight>()
        val income = availability.netIncomeThisMonth
        val available = availability.available

        if (available.isNegative) {
            out.add(Insight.Overspending(-available))
        } else if (income.isPositive) {
            val pct = available.percentOf(income)
            if (pct < 10) out.add(Insight.TightBudget(available, pct))
            else if (pct >= 20) out.add(Insight.HealthySurplus(available, pct))
        }

        val totalFixed = fixedCosts.map { it.monthlyAmount }.total()
        if (income.isPositive && totalFixed.isPositive) {
            val pct = totalFixed.percentOf(income)
            if (pct >= 50) out.add(Insight.HighFixedBurden(totalFixed, pct))
        }

        val rent = fixedCosts.filter { it.category == FixedExpenseCategory.RENT }.map { it.monthlyAmount }.total()
        if (rent.isPositive && income.isPositive) {
            val pct = rent.percentOf(income)
            if (pct >= 33) out.add(Insight.HighHousing(rent, pct))
        }

        if (fixedCosts.size >= 2) {
            val largest = fixedCosts.maxByOrNull { it.monthlyAmount.rappen }
            if (largest != null && largest.monthlyAmount.isPositive) {
                out.add(Insight.LargestFixedCost(largest.name, largest.monthlyAmount, largest.monthlyAmount.percentOf(totalFixed)))
            }
        }
        if (variableBudgets.size >= 2) {
            val largest = variableBudgets.maxByOrNull { it.plannedAmount.rappen }
            if (largest != null && largest.plannedAmount.isPositive) out.add(Insight.LargestVariable(largest.name, largest.plannedAmount))
        }

        val overdue = bills.filter { BillClassifier.state(it, today) == BillState.OVERDUE }
        if (overdue.isNotEmpty()) out.add(Insight.OverdueBills(overdue.size, overdue.map { it.amount }.total()))

        val monthlySavings = savingsGoals.mapNotNull { it.monthlyContribution }.total()
        if (income.isPositive) {
            if (monthlySavings.isZero && available.isPositive && !available.isZero) out.add(Insight.NoSavings)
            else if (monthlySavings.isPositive) {
                val pct = monthlySavings.percentOf(income)
                if (pct >= 10) out.add(Insight.GoodSavingsRate(monthlySavings, pct))
            }
        }

        if (out.isEmpty()) out.add(Insight.AllHealthy)
        return out.sortedWith(compareBy { it.severity.priority }) // stable
    }
}
